File: app/src/main/java/com/camobap/imgmixer/ImageMixer.java
package com.camobap.imgmixer;

import androidx.annotation.Keep;
import androidx.annotation.NonNull;

import java.io.File;

@Keep
public final class ImageMixer {
    static {
        System.loadLibrary("ndk-shader-mixer");
    }

    public ImageMixer(int width, int height) {
        initialize(width, height);
    }

    public void load(@NonNull File imagePath) {
        load0(imagePath.getAbsolutePath());
    }

    public void convert(@NonNull File outFile) {
        convert0(outFile.getAbsolutePath());
    }

    @Keep
    public native void clear();

    @Keep
    private native void convert0(@NonNull String outPath);

    @Keep
    private native void load0(@NonNull String imagePath);

    @Keep
    protected native void initialize(int width, int height);

    @Keep
    protected native void finalize() throws Throwable;

    @Keep
    private long nativeHandle;
}

File: app/src/main/java/com/camobap/imgmixer/ImageMixer.kt
package com.camobap.imgmixer

import androidx.annotation.Keep

import java.io.File

@Keep
class ImageMixer(width: Int, height: Int) {

    @Keep
    private var nativeHandle: Long = 0

    init {
        initialize(width, height)
    }

    fun load(imagePath: File) {
        load0(imagePath.absolutePath)
    }

    fun convert(outFile: File) {
        convert0(outFile.absolutePath)
    }

    @Keep
    external fun clear()

    @Keep
    private external fun convert0(outPath: String)

    @Keep
    private external fun load0(imagePath: String)

    @Keep
    protected external fun initialize(width: Int, height: Int)

    @Keep
    @Throws(Throwable::class)
    protected external fun finalize()

    companion object {
        init {
            System.loadLibrary("ndk-shader-mixer")
        }
    }
}
